Modernize Configuration Infrastructure

fiendie opened the pull request:
Storage Architecture Migration:
    - Replace EEPROM emulation layer with LittleFS JSON file storage
    - Migrate from SysPara functions to modern Config.h class with JSON persistence
    - Add hierarchical configuration structure with automatic validation

    Parameter System Refactor:
    - Replace SysPara with Parameter.h implementing type-safe Parameter classes
    - Support int, uint8, double, float, and string parameter types with validation
    - Implement automatic getter/setter binding with constraint enforcement
  
    Parameter Registry:
    - Replace editableVars with centralized ParameterRegistry system
    - Enable automatic config synchronization when parameters change
    - Provide unified interface for MQTT, web server, and internal parameter access

    Integration Updates:
    - Update MQTT parameter handling to use ParameterRegistry
    - Modify web server endpoints to use new parameter system

    Website Updates:
    - Add MQTT and System sections to parameter page 
    - Add Factory Reset on about page
    - Add config backup and restore feature to about page
    - Add restart button to about page

The main goal of this PR is to remove compile-time dependencies and make everything, including hardware-related settings, configurable at runtime. All config options are stored and read from config.json. Defaults and constraints are defined in the ConfigDef structure and defaults are stored in a map with the complete JSON path as the key, e.g. "hardware.switches.brew.enabled". ParameterRegistry serves as an additional layer on top of Config for web-configurable parameters. It adds information like section, help test, display name, etc. We now have a single point of truth for every setting.

Hardware-related settings are currently not web-configurable. We might add a separate hardware page to the web interface at some point but for the time being the onboarding process would involve the following steps: 
- Flash firmware and filesystem image (e.g. with ESP Flash Download Tool)
- Download default config.json from about page 
- Edit hardware-settings (leds, switches, sensors, etc.)
- Upload config.json
- Edit parameters (pid settings, setpoint, MQTT, etc.)

fiendie committed:
Modernize Configuration Infrastructure

Storage Architecture Migration:
- Replace EEPROM emulation layer with LittleFS JSON file storage
- Migrate from SysPara functions to modern Config.h class with JSON persistence
- Add hierarchical configuration structure with automatic validation

Parameter System Refactor:
- Replace SysPara with Parameter.h implementing type-safe Parameter classes
- Support int, uint8, double, float, and string parameter types with validation
- Implement automatic getter/setter binding with constraint enforcement

Parameter Registry:
- Replace editableVars with centralized ParameterRegistry system
- Enable automatic config synchronization when parameters change
- Provide unified interface for MQTT, web server, and internal parameter access

Integration Updates:
- Update MQTT parameter handling to use ParameterRegistry
- Modify web server endpoints to use new parameter system

Move MQTT settings from userConfig.h to new config infrastructure

scottcondie24 commented:
I tweaked my config.h file to

#include "userConfig.h"

            // MQTT
            _doc["mqtt"]["enabled"] = FEATURE_MQTT;
            _doc["mqtt"]["broker"] = MQTT_SERVER_IP;
            _doc["mqtt"]["port"] = MQTT_SERVER_PORT;
            _doc["mqtt"]["username"] = MQTT_USERNAME;
            _doc["mqtt"]["password"] = MQTT_PASSWORD;
            _doc["mqtt"]["topic"] = MQTT_TOPIC_PREFIX;
            _doc["mqtt"]["hassio"]["enabled"] = MQTT_HASSIO_SUPPORT;
            _doc["mqtt"]["hassio"]["prefix"] = MQTT_HASSIO_DISCOVERY_PREFIX;

I reactivated the LEDs that were commented out with no issues. I am able to download the config.json by navigating to silvia.local/download/config

fiendie committed:
Revamped About page, removed footer fragment to save some heap space

fiendie committed:
Added upload function for config.json backups

* Use proper bool values in config.json
* Add a upload function to About page
* Validate config structure and parameter ranges
* Save config if validation passes and reboot in either case

fiendie committed:
Move EMA_FACTOR, HOSTNAME and OTAPASS to config.json

fiendie committed:
Add endpoint and corresponding button for restarting the machine

fiendie committed:
Add dedicated Parameter constructors for numerical, string and bool values

fiendie committed:
Introduce new Parameter type kEnum, make Log Level web-configurable

fiendie committed:
Made display templates web-configurable

* Added a config parameter for the display template
* Added displayTemplateManager.h with namespaces for each display template
* Removed upgright template for the time being to simplify initialization

fiendie committed:
Move all hardware-related settings from userConfig.h to Config.h/config.json

fiendie committed:
Fix Bug where display template wouldn't be saved correctly

* Fixes a little copy-paste error where choosing the display template
  would overwrite the log level instead
* Move scale parameters to hardware section of the config file for
  consistencFix Bug where display template wouldn't be saved correctly

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+#include <ArduinoJson.h>
+#include <LittleFS.h>
+
+class Config {
```
On discord you discussed namespaces - which I personally think would be great to introduce - but I don't know what the status is wrt. to that since I have just started reading the PR changes. My suggestion would be to do that in a separate PR since adding a namespace (and its indentation) will generate a full-file diff.

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+         *
+         * @return true if successful, false otherwise
+         */
+        bool begin() {
```
I think I would move the initialization to the constructor. In the `main.cpp` you would then just keep a pointer around (`std::shared_ptr<Config> config`) and create the object when due `config = std::make_shared<Config>()` which initializes itself and loads defaults.

The benefit is that you cannot interact with the object without being initialized - something that can happen right now.

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+        /**
+         * @brief Create a new configuration with default values
+         */
+        void createDefaults() {
```
this is only called by `begin()` and should not be available to any other class dealing with a `Config` object, make it `private` (and possibly rename to `create_defaults()` to indicate the "private-ness".

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+
+        // WiFi credentials flag
+        bool getWifiCredentialsSaved() {
+            return _doc["wifi"]["credentials_saved"] | false;
```
I presume they are saved elsewhere (by the wifi manager?)?

fiendie replied:
This is a bit of a relic. I kept it in place because some of the business logic in setup() was relying on it. Can probably be removed!

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+        }
+
+        void setUsePonM(const bool value) {
+            _doc["pid"]["use_ponm"] = value;
```
An even cleaner solution than this would be to have a `PidConfig` class which has `getUsePonM()`, and `Config::getPidConfig()`

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+            return _doc["mqtt"]["hassio"]["prefix"];
+        }
+
+        void setMqttHassioPrefix(const String& value) {
```
I am wondering if a more generic

```cpp
class Config {
  template <typename T>
  T get(const String& topic, const String& key) const;

  template <typename T>
  void set(const String& topic, const String& key, std::variant<String, bool, ...> , T& value);
  }
```

would not make this a cleaner interface

simonspa replied:
Usage:

```cpp

auto myvalue = config.get<String>("mqtt","hassio_prefix");
myvalue += "_something";
config.set("mqtt", "hassio_prefix", myvalue);
```

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+
+        StaticJsonDocument<4096> _doc;
+
+        bool validateAndApplyConfig(const JsonDocument& doc) {
```
This file is *really* long and for the sake of legibility I would separate header and implementation. That makes it a lot easier to understand the structure of the class.

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+                setPidEnabled(doc["pid"]["enabled"].as<bool>());
+            }
+
+            if (doc["pid"].containsKey("use_ponm")) {
```
If I understand the code correctly, the JSON file is only ever accessed/created/altered through this class. IMHO it would then suffice to validate the parameters when setting and reading them instead of cycling through the entire file and re-implementing checks.

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+            if (doc["pid"].containsKey("ema_factor")) {
+                const double value = doc["pid"]["ema_factor"].as<double>();
+
+                if (!validateParameterRange("pid.ema_factor", value, PID_EMA_FACTOR_MIN, PID_EMA_FACTOR_MAX)) {
```
Example here, you use `constrain(/* ... */)` above and here again validate the constrained range - I think this is a bit duplicated

simonspa commented on src/Config.h:
```diff
@@ -0,0 +1,1488 @@
+                }
+
+                setFeatureHeatingLogo(doc["display"]["heating_logo"].as<bool>());
+            }
```
I am trying to understand what this function for example does. There are two options:

* `doc["display"]["heating_logo"]` **is** a boolean: The first condition returns false, function continues. We read the value of `doc["display"]["heating_logo"]` and assign it to `doc["display"]["heating_logo"]`. This is a no-op...?
* `doc["display"]["heating_logo"]` **is not** a boolean: first condition kicks in, function aborts, value stays a non-boolean JSON value.

What am I missing?

simonspa replied:
oh, `doc` and `_doc`, now I see. Okay, still this might be a bit of a cumbersome validation to do for every parameter - and if the input is not correct, the value will not be set.

simonspa commented on src/Parameter.h:
```diff
@@ -0,0 +1,379 @@
+            return _section;
+        }
+
+        [[nodiscard]] int getPosition() const {
```
I like the `[[nodiscard]]` sugar :)

simonspa commented on src/Parameter.h:
```diff
@@ -0,0 +1,379 @@
+
+class Parameter {
+    public:
+        Parameter(
```
I have not followed closely in the past months, but what has become of the idea to us a `class Parameter : public std::variant<std::int32_t, double, String, ...>`? I think it would implement some of the features already that this class provides

simonspa commented on src/ParameterRegistry.cpp:
```diff
@@ -0,0 +1,600 @@
+
+ParameterRegistry ParameterRegistry::_singleton;
+
+// Global variables, needed for backwards compatibility
```
Backwards to what? Older versions? How would they interfere here?
Backwards to still-existing legacy code in `main`?

fiendie replied:
The latter, yes. We should probably move away from global variables in the long run.

simonspa commented on src/ParameterRegistry.cpp:
```diff
@@ -0,0 +1,600 @@
+    // PID Section
+    addParam(std::make_shared<Parameter>(
+        "PID_ON", "Enable PID Controller", kUInt8, sPIDSection, 1,
+        [&config]() -> bool {
```
see comment in `Config` class. If we don;t re-implement setter and getter for individual keys there, using the parameter name (`PID_ON`) as the key it would be much simpler to access them without providing lambdas for accessing every single one of them

simonspa commented on src/ParameterRegistry.cpp:
```diff
@@ -0,0 +1,600 @@
+        },
+        true, "Use PonM mode (<a href='http://brettbeauregard.com/blog/2017/06/introducing-proportional-on-measurement/' target='_blank'>details</a>)", [] { return true; }, &usePonM));
+
+    addParam(std::make_shared<Parameter>(
```
I would suggest changing the `addParam` to accept variadic arguments and do the `make_shared` inside the function:

```cpp
template <typename... Args>
void addParam(Args&&... args) {
  _parameters.emplace_back(std::forward<Args>(args)...));
}
```

simonspa commented on src/ParameterRegistry.cpp:
```diff
@@ -0,0 +1,600 @@
+    // Add all parameters
+    auto addParam = [this](const std::shared_ptr<Parameter>& param) {
+        _parameters.push_back(param);
+        _parameterMap[param->getId()] = param;
```
Do we need to store the parameter pointer twice?

simonspa replied:
The `_parameters` vector is only returned in `getParameters()` but we could just build a vector from the map and return it.

simonspa commented on src/ParameterRegistry.h:
```diff
@@ -0,0 +1,216 @@
+#include <memory>
+#include <vector>
+
+enum ParameterSection {
```
a parameter section seems to have an ID and a name - for me that woukld be enough to make it a `class` or at least ` struct`

simonspa reviewed (commented):
Alright, this is an enormous PR so I'm reading it in chunks. Here are my comments concerning the `Config`, `Parameter` and `ParameterRegistry` classes.

in my opinion, we should think this through a bit more on a architectural level. Right now, for my taste there are too many places in which we have specific setter and getter functions that need to know about value types and parameter names. Ideally, there is only one class that I ask about a variable and all the access code in registries and so on is independent of the type. See some ideas inline.

simonspa commented on src/display/displayTemplateManager.h:
```diff
@@ -0,0 +1,62 @@
+#include "displayCommon.h"
+
+namespace Templates {
+namespace Standard {
```
I would like to put the namespaces into the respective headers instead of wrapping the `#include`.

simonspa replied:
```cpp
namespace Templates::Standard {
    /* ... */
}
```

should work

simonspa replied:
Namespace indentation you should get with 

```diff
diff --git a/.clang-format b/.clang-format
index 8dd84fa..a2b2426 100644
--- a/.clang-format
+++ b/.clang-format
@@ -32,6 +32,7 @@ IndentCaseLabels: true
 IndentCaseBlocks: true
 IndentRequires:  false
 IndentWidth:     4
+NamespaceIndentation: All
 PackConstructorInitializers: NextLineOnly
 PointerAlignment: Left
 QualifierAlignment: Leave
```

To apply this to all files in the repo, run

```
$ pre-commit run --all-files
```

otherwise the new format will only be applied to the changed files.

simonspa replied:
The reason it didn't work here previously likely is that `#include` is no code but a pre-processor macro, and they are never indented.

simonspa commented on src/display/displayTemplateManager.h:
```diff
@@ -0,0 +1,62 @@
+
+class DisplayTemplateManager {
+    public:
+        enum Template {
```
Hm, What I would do is not use namespaces but classes. You have a base `class Template` and then a `class Standard : public Template`. Then the manager instance does not need an additional `enum` but can directly say `currentPrintScreen = std::make_shared<T>()`

simonspa commented on src/embeddedWebserver.h:
```diff
@@ -86,72 +52,91 @@ String getTempString() {
-            default:
-                return F("Unknown type");
-                break;
+        const std::shared_ptr<Parameter> e = ParameterRegistry::getInstance().getParameterById(varName.c_str());
```
```suggestion
        const auto e = ParameterRegistry::getInstance().getParameterById(varName.c_str());
```

You don't have to type your fingers to the bone. :wink:

simonspa commented on src/hardware/StandardLED.cpp:
```diff
@@ -7,18 +7,16 @@
+        gpio.write(state ? LOW : HIGH); // Inverted logic
+    }
+    else {
+        gpio.write(state ? HIGH : LOW); // Normal logic
```
This entire block should be equivalent to a XOR

```suggestion
        gpio.write(state != inverted ? HIGH : LOW);
```

simonspa commented on src/mqtt.h:
```diff
@@ -7,53 +7,75 @@
+        return;
+    }
+
+    mqtt_enabled = registry.getParameterById("MQTT_ENABLED")->getBoolValue();
```
should have noted above, but you could template the function directly to make this

```cpp
    mqtt_enabled = registry.get<bool>("MQTT_ENABLED");
```

simonspa reviewed (commented):
Alright, that's the rest - overall really nice cleanup! :rocket:

fiendie committed:
Remove remaining dependencies on userConfig.h

fiendie committed:
Split up Parameter into source and header files

fiendie committed:
Fix issue where language wouldn't get read from config

* Fixes a bug where the display language wouldn't be properly read and
  set from config.json
* Add display language to web-configurable parameters

scottcondie24 committed:
removed FEATURE_ references, set TARE_ON and CALIBRATE_ON to return true so they show in main screen

fiendie committed:
Use JSON path as ID in ParameterRegistry, fix config upload

* Instead of unique IDs for Parameters, use the JSON path, e.g.
  "pid.enabled" instead of "PID_ON"
* Added missing OLED parameters to config upload validation
* Added user feedback on the About page when validation fails

fiendie committed:
Fix toggles and temperature input on home page

fiendie committed:
Implement template getter method for the Parameter and template setter method for ParameterRegistry

fiendie committed:
Introduce ConfigDef struct to map JSON config entries to their constraints

fiendie committed:
Refactored Config.h to get rid of hard-coded getters and setters

Added factory reset button on about page

fiendie committed:
Add convenience methods in ParameterRegistry to make adding new Parameters more straight-forward and reduce boilerplate

fiendie committed:
Avoid temporary strings in template getter and setter methods to prevent heap corruption

Tweak visibility settings of brew-related parameters if brew switch is enabled/disabled

fiendie committed:
Add config parameter reference

fiendie committed:
Add python script to create binary flash package

fiendie committed:
Prettify config.json for download

fiendie committed:
Moved system functions from About to separate page
